feat: add joystick

pavelskipenes opened the pull request:
- rewrite the task switching logic
- implement joystick tasks
- add default emulated task (not working until a joystick is detected. cc: @incebellipipo @Andeshog)
- bug: members in `JoystickButtons` are not initialized but they are in `JoystickAxes` which is identical in all other ways. Initial values has been set
- bug: `timer_callback()` continuously resets the Joystick task to "simple" from somewhere cc: @incebellipipo @Andeshog.

Needs testing with thruster allocation since all tasks besides simple sends commands to thruster allocation. simple task seem to work fine

cc: @AuroraPuffin @lordAudun @marcoomoretti 
clean setup:
```bash
git clone ...
git checkout chaney_thrust_allocation
git checkout pavel_joystick
git submodule init
git submodule update
```

to check out and test the code
```bash
git stash
git fetch
git checkout pavel_joystick
cd ~/ros_ws/
colcon build --symlink-install
ros2 launch tmr4243_utilities utilities.simulation.launch.py
ros2 launch template_joystick_control joystick.launch.py
ros2 launch template_thrust_allocation thrust_allocation.launch.py
# to test with thrust allocation
git merge --squash chaney_thrust_allocation
```
running in hardware mode
```bash
ros2 launch tmr4243_utilities utilities.physical.launch.py
systemctl restart docker
ros2 service call /voyager/thruster/enable std_srvs/srv/Empty {}
```

restore consistent state after testing:
```bash
git restore --staged template_thrust_allocation/
git checkout -
cd ~/ros_ws/
colcon build --symlink-install
```

pavelskipenes committed:
config: map joystick inputs

pavelskipenes committed:
joystick: dispaly raw joystick values

- rename u to tau
- formatting
- display raw joystick values
- use assert instead of if statement

pavelskipenes committed:
joystick: mapping axes

In this commit the `JoystickMapping` class has been split into
`JoystickAxes` and `JoystickButtons` to logically separate logically
separate enteties. One of the axes index is not correct accordingo
to the official ROS documentation. Tested with ros package `joy` and
`joy_linux`.

In this commit the axes has been mapped using a playstation 4
controller. Button mapping have been mapped from the official
documentation but not yet tested with hardware.

pavelskipenes committed:
joystick: remap inputs for PS4

pavelskipenes committed:
joystick: refactor

- add simulated joystick
- define topics where to publish results

pavelskipenes committed:
joystick: update import references

pavelskipenes committed:
joystick: first working version

- imports: fix inconsistencies
- logging: logs new task on change not continiously
- fix/emulator: does no longer return NaN for yaw
- emulator: tuning
- static analysis: add mypy config
- type hinting: add more type hints, prefer ArrayLike over ndarray
- fix: bug where setting a new task immediatly resets it to default

There is currently a bug where button mappings have to be hard coded on
the enum class. I don't know why that is the case, but this is not needed
on the JoystickAxes class. Current workaround has been to set those
values on the class and not accept them from any other sources.

pavelskipenes committed:
joystick: set default task to emulated

set the default joystick task to emulated. Switching tasks required a
joystick so no other default makes sense.

pavelskipenes committed:
joystick: rename file names

pavelskipenes committed:
joystick: remove debugging dependency

pavelskipenes commented:
update:
pipeline joystick -> thrust allocation works.
resulting actuation does not make sense tho in body.

Andeshog commented on template_joystick_control/template_joystick_control/control_node.py:
```diff
@@ -0,0 +1,154 @@
+            )
+        )
+
+        # self.timer = self.create_timer(0.1, self.timer_callback)
+
+    def timer_callback(self) -> None:
+        # This shit is setting the task to simple overwriting my other shit
+        # for some reason something somewhere is spamming the task to be 'simple'
+        pass
+        # maybe_new_task_name = self.get_parameter(
+        #     'task').get_parameter_value().string_value
+        # match maybe_new_task_name:
+        #     case str():
+        #         try:
+        #             self.get_logger().info(
+        #                 f"[Joystick] crazy shit is being set: {maybe_new_task_name}")
+        #             self.task = Task(maybe_new_task_name)
+        #         except ValueError:
+        #             self.get_logger().error(
+        #                 f"[Joystick] Can not construct task name from {maybe_new_task_name} because it's not in {list(Task._member_names_)}. Request Ignored")
+        #     case not_a_string:
+        #         self.get_logger().error(
+        #             f"[Joystick] received an unknown variable {not_a_string}")
+        #         raise ValueError(not_a_string)
+        #
+        # self.get_logger().info(
+        #     f"[Joystick] Parameter task: {self.task}", throttle_duration_sec=1.0)
```
There is no reason to keep this `timer_callback` btw. What you should do instead is add 
```python
self.task = self.get_parameter('task').value
```
at the end of `__init__`. Then in `joy_callback` this value will be overwritten by the new one

pavelskipenes replied:
ahh nice

pavelskipenes replied:
tested and fixed in c77bc500a468824ae9abe2411ac685253490dd1b and 3d0015eeb523f9351e5343d106b2d1fc5c6275d3

pavelskipenes commented on template_joystick_control/template_joystick_control/task_simple.py:
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env python3
+
```
I'll not take blame for that one 😅
`git blame template_joystick_force/template_joystick_force/joystick_force_basin_relative.py`
```
eb4d99ee template_joystick_force/template_joystick_force/joystick_force_basin_relative.py     (Emir Cem Gezer 2024-02-05 02:21:01 +0100  1) #!/usr/bin/env python3
```

pavelskipenes replied:
fixed in 067ddb5

pavelskipenes committed:
feat/joystick: add inital task

pavelskipenes committed:
fix/joystick: handle all unexpected edgecases

pavelskipenes committed:
fix/joystick: remove shebang from non exacutables

pavelskipenes committed:
feat/joystick: add error handling for basin mode

- format imports

pavelskipenes committed:
fix/joystick: transpose the rotation matrix

The rotation matrix needs to be transposed to rotate from NED
to body since the commands are interpreted in body frame.

pavelskipenes committed:
refactor/joystick: refactor error handling

Error handling has been tested by hardcoding the output from basin

pavelskipenes committed:
joystick: remove emulated joystick mode

Tasks return command when they receive a joystick message. That means
the joystick emulator, which was intended as a replacement for users
without a physical joystick, depends on there being a joystick present.
This does not make sense so this "feature" was removed.

pavelskipenes committed:
refactor/joystick: rename channel to topic

topic is more appropriate ros term but means the same thing

pavelskipenes committed:
refactor/joystick: sort imports

pavelskipenes committed:
refactor/joystick: remove todo, improve type hinting

- `.flatten()` is needed because tasks return the actuation in vector
form for consistency with math notation.
- use `Literal` when a subset of a type is ever returned.

pavelskipenes committed:
refactor/joystick: define distinct types for u and tau

Actuation is u
Command is tau

pavelskipenes committed:
refactor/joystick: error handling

Refactor the code so that `task_*` functions will always return a
valid value and never any error. Hense remove the `Error?  type.
Values that can be invalid are checked at the receiving end before
they're saved and commited to the `Joystick` object.

- rename rotation matrix to indicate that it's inverted.

- reduce amount of code by simplifying `task_body()`.

pavelskipenes committed:
fix/joystick: update accepted heading range

pavelskipenes committed:
add/joystick: add kinematics

nautamat reviewed (approved):
LGTM

pavelskipenes commented:
waiting for merge of #2 before rebasing and mergin this PR.

lordAudun reviewed (approved):
Slayyy skibidi
